### ISSUE TITLE
[HOLD for payment 2023-12-12] [$1000] Chat - The green line is displayed chaotically at the recipient

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#27456]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to https://staging.new.expensify.com/ on an Android (User B)
2. Sign into Android app (User A)
3. Send a message from User B to User A
4. In LHN of user A, the message received will be shown in bold
5.Open the message
6.From user B, send multiple messages to User A

## Expected Result:
As user A is already in conversation page, the green line must not be shown when messages received from user B

## Actual Result:
The green line is displayed chaotically at the recipient

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.70-2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/1af8a7f0-1afc-4101-bb8f-26f3fd523695

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018a50aa27edf905a3</li>
        <li>Upwork Job ID: 1703890402212745216</li>
        <li>Last Price Increase: 2023-11-11</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Contributor | 27473899</li>
</ul></li>
    </ul>
</details>

**Comment by Christinadobrzyn** [comment]
I can't reproduce this going to reach back out to QA to see if they can confirm it's still happening and update the steps to reproduce.

**Comment by lanitochka17** [comment]
Issue reproducible on Build 1.3.70-5
SG A50/Android11
Messages were sent one after another

https://github.com/Expensify/App/assets/78819774/dd41cfd0-e78d-4391-ad7e-3978d817ffe7

**Comment by Christinadobrzyn** [comment]
I still can't reproduce this... there's no green line in a chat

Using V1.3.71-2 on the Android app app Samsung Galaxy S22
Using v1.3.7 0-8 on web app Samsung 

@lanitochka17 does something need to be added to these steps to make this happen?

**Comment by Christinadobrzyn** [comment]
hey @mananjadhav can you give me your thoughts on https://github.com/Expensify/App/issues/27456#issuecomment-1724503801

**Comment by Christinadobrzyn** [comment]
nudge @mananjadhav on https://github.com/Expensify/App/issues/27456#issuecomment-1724503801

**Comment by mananjadhav** [comment]
I tried a few times, but I couldn't. I checked on the emulator. Does it require a physical device to check?

**Comment by Christinadobrzyn** [comment]
I was using the Chrome Website on a physical Android device (User B) and messaging the Android app through [Browserstack](https://app-live.browserstack.com/#os=android&os_version=13.0&device=Google+Pixel+7&zoom_to_fit=true&full_screen=true&speed=1) (User A).

**Comment by Christinadobrzyn** [comment]
I updated the OP to see if that helps @mananjadhav - can you try again to reproduce and let me know what you think is best next steps here... maybe we should close since it's difficult to reproduce?

**Comment by mananjadhav** [comment]
@Christinadobrzyn I think you can reassign this one. I haven't been able to spend as much time as this requires.

**Comment by Christinadobrzyn** [comment]
Hey @Santhosh-Sellavel! Catching you up here, we're still reproducing this and seeing if it's necessary for a fix based on this [comment](https://github.com/Expensify/App/issues/27456#issuecomment-1724503801). Let me know what you think!

**Comment by Santhosh-Sellavel** [comment]
Yeah this is something we should fix, if its reproducible.

**Comment by Christinadobrzyn** [comment]
Thanks @Santhosh-Sellavel can this be external?

**Comment by Santhosh-Sellavel** [comment]
Yes, it can be external.

**Comment by Christinadobrzyn** [comment]
still waiting on proposals. Do you think this would be something that would fit one of our contractors @Santhosh-Sellavel? I can reach out to SWM to see if they have the bandwidth?

**Comment by Santhosh-Sellavel** [comment]
That's good idea, but lets wait for few days or a week. If no takers by then let's get help from SWM or Callstack!

**Comment by roksanaz** [comment]
I am Roksana from Callstack - expert contributor group. I’d like to work on this job.

**Comment by Santhosh-Sellavel** [comment]
> I am Roksana from Callstack - expert contributor group. I’d like to work on this job.

Can you submit a formal proposal please, thanks!

**Comment by mountiny** [comment]
Assigning @roksanaz But still please post a proposal here to discuss before raising a PR thanks!

**Comment by roksanaz** [comment]
I'm working on a proposal (should be ready tomorrow), but just to be sure: the desirable behavior is that when the user is active in chat screen, there shouldn't be a green line for new messages, am I correct? (citing the expected result: "As user A is already in conversation page, the green line must not be shown when messages received from user B")

**Comment by Santhosh-Sellavel** [comment]
Let's consider a scenario to understand the expected outcome. 

There a two User A & User B. Following or the sequence of actions 

1. `User A` is chatting with someone else and new messages are received from `User B`, thus making `User B` in Sidebar LHN bold.
2. Now `User A` navigates to `User B's` chat. 
**Expected Behaviour is a Line indicator that should be shown above the first unread message.**
3. User B still sends a couple more messages. 
**Expected Behaviour is a Line indicator that should persist it's position**
4. If User A responds or Navigates to a different change and comes back to chat then the 
**Expected Behaviour is that the indicator should be hidden now.**
5.` User A` is active on User B's chat and scrolling the old chat, now new messages are received. **Now the expected behavior is when the user scrolls to the new message manually or by using the new messages badge, should see an indicator above the first unread item.**
6. ` User A` is active on User B's chat but viewing the latest, now new messages are received.
**I'm not sure, but I guess there should not be any indicator new messages should be shown right away.**


@Christinadobrzyn/@mountiny Can you loop someone from the product team to confirm the expected behavior, correct me if wrong on the above. thanks!

**Comment by roksanaz** [comment]
I'm sorry for taking so much time. We had a discussion that maybe another PR concerning unread messages might resolve this issue, but unfortunately that's not the case :( I need to consult 1 thing and my proposal should be ready after that.

**Comment by Christinadobrzyn** [comment]
I _think_ this is correct but gonna ask Vit if they can double-check- https://github.com/Expensify/App/issues/27456#issuecomment-1758212639

**Comment by Christinadobrzyn** [comment]
Sounds like this job was discussed between @MonilBhavsar and @roksanaz - https://expensify.slack.com/archives/C03UK30EA1Z/p1697013288899459

so I think they should be able to answer your questions @Santhosh-Sellavel once they get things narrowed down?

**Comment by roksanaz** [comment]
@Santhosh-Sellavel and @Christinadobrzyn yes, we discussed the logic from [this comment](https://github.com/Expensify/App/issues/27456#issuecomment-1758212639) and points 1-4 are tested and working in this [PR](https://github.com/Expensify/App/pull/25935), point 5 will be implemented within this issue. 

I'm on the right track with my proposal, should be ready on Monday.

**Comment by roksanaz** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

If the user A is already active in the chat thread with the user B, the user A receives a new message from the user B, a green line appears above the new message. If more messages come (especially at the same time) the green line appears erratically. The expected behaviour is that no green line appears for new messages if the thread is scrolled to the latest message.

### What is the root cause of that problem?

I could reproduce this behaviour for all platforms. The problem isn’t specific only for Android.
The root cause for the problem is that we don’t take into account the position of the user (the vertical offset) in the conditions for the flag called `shouldDisplayNewMarker`, which if it’s true - the green line appears. 
Currently on the first render of the message `shouldDisplayNewMarker` is true if the flag `lastReadTime` sent by the server is smaller than the message’s created time (and it is always true for new messages):

https://github.com/Expensify/App/blob/7d08e237dad360573c113d5f8c34d2b42ef49a93/src/pages/home/report/ReportActionsList.js#L292

Then, the application sends information to the server that the message is read, the message is re-rendered, `shouldDisplayNewMarker` is false and the green line disappears.

### What changes do you think we should make in order to solve the problem?

As the desired behaviour is not to display the green line for any new messages provided that the thread is scrolled to the latest message, we should take into account the vertical offset of the user as it’s the case with the floating button. `shouldDisplayNewMarker` should be false for threads at the end of the threshold before sending the information to the server to avoid green line appearing (or flickering). If the thread is scrolled up to older messages, then the floating button should appear and after scrolling down to the newest messages, the green line should appear above the new message.
There is already a variable in the code called `canDisplayMarker`, which holds the information about the user’s position and if the user is active:

https://github.com/Expensify/App/blob/7d08e237dad360573c113d5f8c34d2b42ef49a93/src/pages/home/report/ReportActionsList.js#L297

But now it’s introduced after `shouldDisplayNewMarker` is already true and taken into account only to decide which message should be marked as new. So my solution is to introduce `canDisplayMarker` earlier and check it while deciding for the first time if `shouldDisplayNewMarker` should be true, like so:   

`shouldDisplayNewMarker = isCurrentMessageUnread && !isMessageUnread(nextMessage, report.lastReadTime) && canDisplayMarker;`

**Comment by Santhosh-Sellavel** [comment]
Looks good, let's hold for the other [PR](https://github.com/Expensify/App/pull/25935) to get merged.

**Comment by MonilBhavsar** [comment]
I think we can move forward here since it doesn't overlap with the other PR. @roksanaz do you think we can get rid of `canDisplayMarker` completely and have only one variable  `shouldDisplayMarker`. We previously thought of doing this, but didn't, but figured it was something we could do. I think this is good opportunity to get rid of two similar variables. Also cc @gedu for thoughts on the above proposal

**Comment by roksanaz** [comment]
@MonilBhavsar We could get rid of this variable, but then `shouldDisplayMarker` would consist of an unnamed condition instead of `canDisplayMarker`, like so:

```js
shouldDisplayNewMarker = isCurrentMessageUnread && 
      !isMessageUnread(nextMessage, report.lastReadTime) &&
      scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD ? reportAction.created < userActiveSince.current : true;
```

 If I were to do a complete refactor of this part, I would make it more readable, so instead of this:
 
https://github.com/Expensify/App/blob/cf16dacf2eb12cd5263e0a1d09eb1638b0fe5c93/src/pages/home/report/ReportActionsList.js#L290-L301
 
 I would write it like that:
 
 ```js
const nextMessage = sortedReportActions[index + 1];
const isCurrentMessageUnread = isMessageUnread(reportAction, report.lastReadTime);
const isNextMessageRead = !isMessageUnread(nextMessage, report.lastReadTime);
const isWithinVisibleThreshold = scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD ? reportAction.created < 
    userActiveSince.current : true;

shouldDisplayNewMarker = isCurrentMessageUnread && isNextMessageRead && isWithinVisibleThreshold;

if (shouldDisplayNewMarker && !messageManuallyMarkedUnread) {
    shouldDisplayNewMarker = reportAction.actorAccountID !== Report.getCurrentUserAccountID();
}

if (shouldDisplayNewMarker) {
    setCurrentUnreadMarker(reportAction.reportActionID);
}
```
Let me know which approach is best for you.

**Comment by MonilBhavsar** [comment]
Looks like last condition is missing  `currentUnreadMarker` . But later one looks clean to me. Thanks!

```JS
 if (!currentUnreadMarker && shouldDisplayNewMarker) {
   setCurrentUnreadMarker(reportAction.reportActionID);
}
```

**Comment by roksanaz** [comment]
I think `!currentUnreadMarker` condition was redundant in line 299, as all of this logic is wrapped in the same condition:

https://github.com/Expensify/App/blob/91e2120e647aa9ee22171b6dbb772b4aebe64992/src/pages/home/report/ReportActionsList.js#L289

So I'll head in this direction with my PR, thank you.

**Comment by MonilBhavsar** [comment]
Yes, you're right 👍 
Thank you!

**Comment by MonilBhavsar** [comment]
Looks like draft PR is up and in being pre-reviewed

**Comment by Santhosh-Sellavel** [comment]
Please assign @situchan as they took over PR while I was away.

**Comment by MonilBhavsar** [comment]
We reverted the PR as it introduced blocker. We need a new PR

**Comment by roksanaz** [comment]
@MonilBhavsar I checked all the issues and I'll create a new PR that fixes this issue without causing https://github.com/Expensify/App/issues/30684 (sorry about that). However, https://github.com/Expensify/App/issues/30696 was not caused by my previous PR, so I didn't address that.

**Comment by roksanaz** [comment]
@MonilBhavsar As I was testing the new logic taking into account the message marked as unread manually (`messageManuallyMarkedUnread`), I stumbled upon a problem.

I found out that this logic:

https://github.com/Expensify/App/blob/9ee3b4a83df14589bd7af42d3aa4aac1179754da/src/pages/home/report/ReportActionsList.js#L193-L206

is fired and `didManuallyMarkReportAsUnread` is `true` even though I don't mark the message as unread by myself. From my investigation it turns out that web app is not affected (tested on mac OS chrome and safari), but native apps and mobile websites are. 

I attach the recording where I simply kept sending the messages from user A from the website (chrome) to user B using Android native application (physical device, Samsung). The new messages were marked as unread and the green line appeared because `messageManuallyMarkedUnread` was `true`. I also attach the debugged logic concerning `messageManuallyMarkedUnread`.


https://github.com/Expensify/App/assets/44186263/d4529e7b-aaa7-42c2-b8d9-f2956f77c1da

<img width="1117" alt="Screenshot 2023-11-02 at 19 01 35" src="https://github.com/Expensify/App/assets/44186263/a0c72105-299e-41a8-b3db-461b57ea9c64">


I have a new branch for this solution (https://github.com/roksanaz/App/tree/27456-30684-green-line), but I can't create a new PR, because the green line bug from this issue won't be  properly solved.

**Comment by roksanaz** [comment]
Hello, I just want to let you know that from now on @artus9033 will take over this issue, so I would be grateful if you could assign this to him :)

**Comment by situchan** [comment]
@artus9033 you may need to comment anything here to be assigned.

**Comment by Christinadobrzyn** [comment]
Can we just get an update here since the payment date has passed? It sounds like we're working on fixing a regression?

**Comment by situchan** [comment]
Original PR was reverted. This needs re-work.

**Comment by situchan** [comment]
Please remove hold for payment in title. I think issue needs to restart from proposal stage.

**Comment by Christinadobrzyn** [comment]
Updated the tile, removed the `awaiting payment` label and added a `help wanted ` label to get proposals

**Comment by artus9033** [comment]
Sure, I'm adding the comment for assignment here

**Comment by Christinadobrzyn** [comment]
I think we're looking for proposals, is that correct @roksanaz and @situchan?

Should we update the GH OP with new steps or are those still the same?

**Comment by situchan** [comment]
Based on https://github.com/Expensify/App/issues/27456#issuecomment-1803762900, @artus9033 is an expert agency and waiting to be assigned here.

**Comment by Christinadobrzyn** [comment]
Awesome! I added @artus9033 to this GH

**Comment by MonilBhavsar** [comment]
@artus9033 @roksanaz @gedu 

Now that we have fixed unread message case in https://github.com/Expensify/App/pull/25935
This proposal https://github.com/Expensify/App/issues/27456#issuecomment-1766129708 works if we only consider it when message is not manually marked as unread

```diff
--- a/src/pages/home/report/ReportActionsList.js
+++ b/src/pages/home/report/ReportActionsList.js
@@ -338,8 +338,9 @@ function ReportActionsList({
                 const nextMessage = sortedReportActions[index + 1];
                 const isCurrentMessageUnread = isMessageUnread(reportAction, lastReadTimeRef.current);
                 shouldDisplay = isCurrentMessageUnread && (!nextMessage || !isMessageUnread(nextMessage, lastReadTimeRef.current));
-                if (!messageManuallyMarkedUnread) {
-                    shouldDisplay = shouldDisplay && reportAction.actorAccountID !== Report.getCurrentUserAccountID();
+                if (shouldDisplay && !messageManuallyMarkedUnread) {
+                    const isWithinVisibleThreshold = scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD ? reportAction.created < userActiveSince.current : true;
+                    shouldDisplay = reportAction.actorAccountID !== Report.getCurrentUserAccountID() && isWithinVisibleThreshold;
                 }
```

**Comment by gedu** [comment]
Looks good from my side

**Comment by artus9033** [comment]
Perfect, thanks @gedu and @MonilBhavsar. It works as intended for me as well. Then, I will put up a PR for you @MonilBhavsar to review.

**Comment by situchan** [comment]
Hope this won't happen but I think we can wait until https://github.com/Expensify/App/pull/25935 hit production in case QA team or any others find any unexpected regression.

**Comment by Christinadobrzyn** [comment]
Following - https://github.com/Expensify/App/pull/25935#issuecomment-1804272308

**Comment by MonilBhavsar** [comment]
I think we can test it appropriately and get this PR out. Let's hold onto that

**Comment by Christinadobrzyn** [comment]
I'm going ooo until Monday Nov 27th so going to assign another BZ teammate to monitor this until I'm back. Thanks!

Status: I'm not 100% sure where we are with this, @MonilBhavsar @situchan can you provide more information?

**Comment by MonilBhavsar** [comment]
Taking this off HOLD, cc @artus9033

**Comment by artus9033** [comment]
Sure thing @MonilBhavsar, then I will be waiting for information when / if https://github.com/Expensify/App/pull/25935 enters production and prepare the PR afterwards, as you pointed above.

**Comment by MonilBhavsar** [comment]
That PR is deployed to staging and will be deployed to production tonight. It has also been checked off successfully in QA test run. I think we don't need to hold onto that. Feel free to create a PR! 🚀

**Comment by artus9033** [comment]
Oh, I see, I misunderstood the last messages - perfect, I will get it done, then.

**Comment by situchan** [comment]
Potential regressions from https://github.com/Expensify/App/pull/25935:
- https://github.com/Expensify/App/issues/31617
- https://github.com/Expensify/App/issues/31621
- https://github.com/Expensify/App/issues/31623

**Comment by artus9033** [comment]
@MonilBhavsar, [the PR](https://github.com/Expensify/App/pull/31627) is ready for review.

**Comment by Christinadobrzyn** [comment]
preparing for payment

Payouts due:

Issue Reporter: NA
Contributor: NA @roksanaz (is not paid in Upwork or NewExpensify)
Contributor+: $1000 @situchan (in Upwork)

Eligible for 50% #urgency bonus? N

Upwork job is [here](https://www.upwork.com/jobs/~018a50aa27edf905a3).

**Comment by Christinadobrzyn** [comment]
Just checking on the payment of this because there were several people working on this. 

@MonilBhavsar @artus9033 and @situchan can you confirm this [payment summary](https://github.com/Expensify/App/issues/27456#issuecomment-1851069929) is correct. 

Do we have a regression test for this?

**Comment by artus9033** [comment]
@Christinadobrzyn it seems like I am listed as a paid contributor, however, I should be a contractor (member of Callstack). Could you update that please?

**Comment by situchan** [comment]
> Just checking on the payment of this because there were several people working on this.
> 
> @MonilBhavsar @artus9033 and @situchan can you confirm this [payment summary](https://github.com/Expensify/App/issues/27456#issuecomment-1851069929) is correct.
> 
> Do we have a regression test for this?

Correct.

I think we already we have regression test case since reported by QA team.

**Comment by MonilBhavsar** [comment]
Yes correct! We can close this once we have issued payment to everyone

**Comment by Christinadobrzyn** [comment]
Awesome! Thanks @situchan and @MonilBhavsar - I paid @situchan based on this payment summary - https://github.com/Expensify/App/issues/27456#issuecomment-1851069929

I think this is good to close but let me know if I'm missing anything!
